test/nlp_expr.jl

```diff
@@ -0,0 +1,200 @@
+    @test string(@expression(model, x + 3 * sin(x))) ==
+          "+(+(x), *(3.0, sin(x)))"
+    @test string(@expression(model, 2 * x + 3 * sin(x))) ==
+          "+(+(0.0, *(2.0, x)), *(3.0, sin(x)))"
+    @test string(@expression(model, x^2 + 3 * sin(x))) ==
+          "+(+(0.0, *(1.0, x, x)), *(3.0, sin(x)))"
+    @test string(@expression(model, sin(x) + 3 * sin(x))) ==
+          "+(sin(x), *(3.0, sin(x)))"
+    @test string(@expression(model, sin(x) + 3 * x)) == "+(sin(x), 3 x)"
+    @test string(@expression(model, sin(x) + 3 * x * x)) ==
+          "+(sin(x), 3 $(x^2))"
+    return
+end
+
+function test_expression_submul()
+    model = Model()
+    @variable(model, x)
+    @test string(@expression(model, x - 3 * sin(x))) ==
+          "-(+(x), *(3.0, sin(x)))"
+    @test string(@expression(model, 2 * x - 3 * sin(x))) ==
+          "-(+(0.0, *(2.0, x)), *(3.0, sin(x)))"
+    @test string(@expression(model, x^2 - 3 * sin(x))) ==
+          "-(+(0.0, *(1.0, x, x)), *(3.0, sin(x)))"
+    @test string(@expression(model, sin(x) - 3 * sin(x))) ==
+          "-(sin(x), *(3.0, sin(x)))"
+    @test string(@expression(model, sin(x) - 3 * x)) == "-(sin(x), 3 x)"
+    @test string(@expression(model, sin(x) - 3 * x * x)) ==
+          "-(sin(x), 3 $(x^2))"
+    return
+end
+
+function test_constraint_lessthan()
+    model = Model()
+    @variable(model, x)
+    @constraint(model, c, 2.0 * sin(x)^2 + cos(x) / x <= 1)
+    nlp = nonlinear_model(model)
+    @test nlp[index(c)] isa MOI.Nonlinear.Constraint
+    @test nlp[index(c)].set == MOI.LessThan(0.0)
+    return
+end
+
+function test_constraint_greaterthan()
+    model = Model()
+    @variable(model, x)
+    @constraint(model, c, 2.0 * sin(x)^2 + cos(x) / x >= 1)
+    nlp = nonlinear_model(model)
+    @test nlp[index(c)] isa MOI.Nonlinear.Constraint
+    @test nlp[index(c)].set == MOI.GreaterThan(0.0)
+    return
+end
+
+function test_constraint_equalto()
+    model = Model()
+    @variable(model, x)
+    @constraint(model, c, 2.0 * sin(x)^2 + cos(x) / x == 1)
+    nlp = nonlinear_model(model)
+    @test nlp[index(c)] isa MOI.Nonlinear.Constraint
+    @test nlp[index(c)].set == MOI.EqualTo(0.0)
+    return
+end
+
+function test_constraint_interval()
+    model = Model()
+    @variable(model, x)
+    @constraint(model, c, 0 <= 2.0 * sin(x)^2 + cos(x) / x <= 1)
+    nlp = nonlinear_model(model)
+    @test nlp[index(c)] isa MOI.Nonlinear.Constraint
+    @test nlp[index(c)].set == MOI.Interval(0.0, 1.0)
+    return
+end
+
+function test_user_defined_function()
+    model = Model()
+    @variable(model, x)
+    f(x::Real) = x^2
+    register(model, :f, 1, f; autodiff = true)
+    g = UserDefinedFunction(model, :f, 1)
+    @test string(@expression(model, g(x))) == "f(x)"
+    @test string(g(x) + g(x)) == "+(f(x), f(x))"
+    return
+end
+
+function test_nonlinear_matrix_algebra()
+    model = Model()
+    @variable(model, X[1:3, 1:3], Symmetric)
+    @objective(model, Max, sum(X^4 .- X^3))
```

Comment:
@ccoffrin this example is also working now ✅ (but it secularizes the expression via operator overloading, so not the most efficient method)